src/scene/graph-node.js

```diff
@@ -589,6 +589,121 @@ class GraphNode extends EventHandler {
         return null;
     }
 
+    /**
+     * Search the graph node and all of its ascendants for the nodes that satisfy some search
+     * criteria.
+     *
+     * @param {FindNodeCallback|string} attr - This can either be a function or a string. If it's a
+     * function, it is executed for each ascendant node to test if node satisfies the search
+     * logic. Returning true from the function will include the node into the results. If it's a
+     * string then it represents the name of a field or a method of the node. If this is the name
+     * of a field then the value passed as the second argument will be checked for equality. If
+     * this is the name of a function then the return value of the function will be checked for
+     * equality against the valued passed as the second argument to this function.
+     * @param {object} [value] - If the first argument (attr) is a property name then this value
+     * will be checked against the value of the property.
+     * @returns {GraphNode[]} The array of graph nodes that match the search criteria.
+     * @example
+     * // Finds all nodes that have a group element component
+     * var groups = element.findInParent(function (node) {
+     *     return node.element && node.element.type === pc.ELEMENTTYPE_GROUP;
+     * });
+     * @example
+     * // Finds all nodes that have the name property set to 'Test'
+     * var entities = entity.findInParent('name', 'Test');
+     */
+    findInParent(attr, value) {
+        let result, results = [];
+
+        if (attr instanceof Function) {
+            const fn = attr;
+
+            result = fn(this);
```

Comment:
`result` has more context than it requires and you could turn it into a `const`, the results array should probably also be a `const` and then simply `results.push(...this._parent.findInParent(fn))`.

(everything is arguable, I just prefer `const` and not creating new intermediate arrays, that makes it easier to argue about)

And it is called `findInParent`, but doesn't this also check the same entity aswell? It would be like referring to `>` as `>=`, which is kinda confusing?

Reply:
I agree about `result` and array spreading, this is mainly copy/pasted from current functions `find` and `findOne`. Maybe those two should be also rewritten then.

It checks the current entity because I'm used to Unity where `GetComponent`, `GetComponentInChildren` and `GetComponentInParent` are all checking the current node. It prevents the need to use a weird syntax or to create a child node only for getting all matching entities, which isn't such an unusual use case imho.
Moreover, `find` and `findOne` are also checking current entity, which can also be confusing and is harder to circumvent if you want to ignore the current entity.